interface/src/checksums_storage.rs

```diff
@@ -0,0 +1,149 @@
+use async_trait::async_trait;
+use solana_sdk::pubkey::Pubkey;
+
+/// Type of checksum for bubblegum epochs and account NFT buckets.
+/// It is technically a SHA3 hash.
+pub type Chksm = [u8; 32];
+
+/// Data transfer object for bubblegum grand epoch.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BbgmGrandEpochCksm {
+    pub tree_pubkey: Pubkey,
+    pub checksum: Option<[u8; 32]>,
+}
+
+/// Data transfer object for bubblegum grand epoch checksums
+/// for listing multiple grand epochs for a tree
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BbgmGrandEpochCksmWithNumber {
+    pub grand_epoch: u16,
+    pub checksum: Option<[u8; 32]>,
+}
+
+/// Data transfer object for bubblegum epoch.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BbgmEpochCksm {
+    pub epoch: u32,
+    pub tree_pubkey: Pubkey,
+    pub checksum: Option<[u8; 32]>,
+}
+
+/// Data transfer object for bubblegum tree change.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct BbgmChangeRecord {
+    pub tree_pubkey: Pubkey,
+    pub slot: u64,
+    pub seq: u64,
+    pub signature: String,
+}
+
+/// Used to specify offset when fetching a portion of
+/// bubblegum tree changes.
+#[derive(Debug, Clone, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BbgmChangePos {
+    pub slot: u64,
+    pub seq: u64,
+}
+
+/// Data transfer object for account NFT grand bucket.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct AccGrandBucketCksm {
+    pub grand_bucket: u16,
+    pub checksum: Option<[u8; 32]>,
+}
+
+/// Data transfer object for account NFT bucket.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct AccBucketCksm {
+    pub bucket: u16,
+    pub checksum: Option<[u8; 32]>,
+}
+
+/// Data transfer object for account NFT last received state.
+#[derive(Debug, Clone, PartialOrd, Ord)]
+pub struct AccLastChange {
+    pub account_pubkey: Pubkey,
+    pub slot: u64,
+    pub write_version: u64,
+    pub data_hash: u64,
```

Comment:
Also, we do not track data_hashes for transaction-based protocol. Maybe we need also them here? I think it makes sense, because it will give us additional 'resist' from Solana forks

Reply:
IMO, we won't gain anything by adding `data_hash` to bubblegum, since we do already have forks cleaner and sequence consistency checks for it.